opensearch-operator/pkg/reconcilers/upgrade.go

```diff
@@ -0,0 +1,363 @@
+package reconcilers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/Masterminds/semver"
+	"github.com/banzaicloud/operator-tools/pkg/reconciler"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	opsterv1 "opensearch.opster.io/api/v1"
+	"opensearch.opster.io/opensearch-gateway/services"
+	"opensearch.opster.io/pkg/builders"
+	"opensearch.opster.io/pkg/helpers"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var (
+	ErrVersionDowngrade = errors.New("version requested is downgrade")
+	ErrMajorVersionJump = errors.New("version request is more than 1 major version ahead")
+	ErrUnexpectedStatus = errors.New("unexpected upgrade status")
+)
+
+type UpgradeReconciler struct {
+	client.Client
+	reconciler.ResourceReconciler
+	ctx               context.Context
+	osClient          *services.OsClusterClient
+	recorder          record.EventRecorder
+	reconcilerContext *ReconcilerContext
+	instance          *opsterv1.OpenSearchCluster
+}
+
+func NewUpgradeReconciler(
+	client client.Client,
+	ctx context.Context,
+	recorder record.EventRecorder,
+	reconcilerContext *ReconcilerContext,
+	instance *opsterv1.OpenSearchCluster,
+	opts ...reconciler.ResourceReconcilerOption,
+) *UpgradeReconciler {
+	return &UpgradeReconciler{
+		Client: client,
+		ResourceReconciler: reconciler.NewReconcilerWith(client,
+			append(opts, reconciler.WithLog(log.FromContext(ctx).WithValues("reconciler", "scaler")))...),
+		ctx:               ctx,
+		recorder:          recorder,
+		reconcilerContext: reconcilerContext,
+		instance:          instance,
+	}
+}
+
+func (r *UpgradeReconciler) Reconcile() (ctrl.Result, error) {
+	// If versions are in sync do nothing
+	if r.instance.Spec.General.Version == r.instance.Status.Version {
+		return ctrl.Result{}, nil
+	}
+
+	lg := log.FromContext(r.ctx)
+
+	// If version validation fails log a warning and do nothing
+	if err := r.validateUpgrade(); err != nil {
+		lg.V(1).Error(err, "version validation failed", "currentVersion", r.instance.Status.Version, "requestedVersion", r.instance.Spec.General.Version)
+		return ctrl.Result{}, err
+	}
+
+	// If there is work to do create an Opensearch Client
+	username, password, err := helpers.UsernameAndPassword(r.ctx, r.Client, r.instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	clusterClient, err := services.NewOsClusterClient(fmt.Sprintf("https://%s.%s:9200", r.instance.Spec.General.ServiceName, r.instance.Namespace), username, password)
```

Comment:
are you deploying new client for communicating with the cluster when the upgrade in progress ?
is it new a nodePool ?

Reply:
Correct, we have a field for it in the struct.  If an upgrade is required we create a new client and add it to the struct state.  That client is then used for the rest of the upgrade.